mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/extractor/impl/ContainerExtractorBinder.java

```diff
@@ -282,15 +287,58 @@ void addCandidate(ExtractorContributor contributor) {
 
 		@Override
 		public boolean tryAppend(ExtractorResolutionState<?> state) {
+			Set<PojoTypeModel<?>> encounteredTypes = new HashSet<>();
+			encounteredTypes.add( state.extractedType );
+			return tryAppend( state, state.extractedType, encounteredTypes );
+		}
+
+		private boolean tryAppend(ExtractorResolutionState<?> state, PojoTypeModel<?> initialType,
+				Set<PojoTypeModel<?>> encounteredTypes) {
 			for ( ExtractorContributor extractorContributor : candidates ) {
 				if ( extractorContributor.tryAppend( state ) ) {
+					if ( !encounteredTypes.add( state.extractedType ) ) {
+						throw log.defaultContainerExtractorCyclicRecursion( initialType, state.extractorNames );
+					}
 					// Recurse as much as possible
 					tryAppend( state );
 					return true;
 				}
 			}
 			return false;
 		}
+
+		public boolean tryMatch(ExtractorResolutionState<?> state, List<String> toMatch) {
```

Comment:
just checking if wether you were planning to pull this into `ExtractorContributor` since the `SingleExtractorContributor` class is private so won't make much difference...

Reply:
Well at the moment I don't have a use case for this method on anything other than the default contributor (`FirstMachingSomethingContributor`), so no, I didn't plan to.

Maybe when there's a reason to :shrug: